### ISSUE TITLE
Gloo throwing "EnforceNotMet" on distributed shutdown when using CUDA

### ISSUE THREAD

**Issue by ArEsKay3** [pytorch/pytorch#2530]
This is showing up when tearing down DistributedDataParallel now that https://github.com/pytorch/pytorch/issues/2520 is fixed. However, it doesn't seem to be related to DistribuedDataParallel code as I can reproduce it with just dist.broadcast(). Simple repro below.

Error:
```
terminate called after throwing an instance of 'gloo::EnforceNotMet'
  what():  [enforce fail at /tmp/pip-fj8uIV-build/torch/lib/gloo/gloo/cuda.cu:249] error == cudaSuccess. 29 vs 0. Error at: /tmp/pip-fj8uIV-build/torch/lib/gloo/gloo/cuda.cu:249: driver shutting down
```

Repro:
```python
from torch import distributed as dist
import torch

dist.init_process_group('gloo', init_method='file://<FILEPATH>', \
    world_size=2)
print('Finished initializing process group.')

print('Broadcasting tensor of ones from rank 0.')
tensor = torch.rand(2, 3).cuda()
rank = dist.get_rank()
if rank == 0:
    tensor = torch.ones(2,3).cuda()
dist.broadcast(tensor, 0)
if rank == 1:
    t2 = tensor.add(20)
    print(t2)
```

**Comment by gitorres** [comment]
I ran this with `gdb --args python main.py`  and got the following stack trace. Although I'm not sure how useful it is:
```
#0  0x00007ffff6a1d428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff6a1f02a in __GI_abort () at abort.c:89
#2  0x00007fff79cfebb5 in __gnu_cxx::__verbose_terminate_handler() () from /home/curiousbot/anaconda/lib/libstdc++.so.6
#3  0x00007fff79cfcd56 in ?? () from /home/curiousbot/anaconda/lib/libstdc++.so.6
#4  0x00007fff79cfbdb9 in ?? () from /home/curiousbot/anaconda/lib/libstdc++.so.6
#5  0x00007fff79cfc9ca in __gxx_personality_v0 () from /home/curiousbot/anaconda/lib/libstdc++.so.6
#6  0x00007ffff499e863 in _Unwind_RaiseException_Phase2 (exc=exc@entry=0x114b090, context=context@entry=0x7fffffffdfb0) at ../.././libgcc/unwind.inc:62
#7  0x00007ffff499ebfb in _Unwind_RaiseException (exc=0x114b090) at ../.././libgcc/unwind.inc:131
#8  0x00007fff79cfcf95 in __cxa_throw () from /home/curiousbot/anaconda/lib/libstdc++.so.6
#9  0x00007fffc9bd9322 in gloo::CudaHostPointer<float>::~CudaHostPointer() () from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#10 0x00007fffc9bc4122 in gloo::CudaBroadcastOneToAll<float, gloo::CudaHostWorkspace<float> >::~CudaBroadcastOneToAll() ()
   from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#11 0x00007fffd5485b69 in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release (this=0x2fa4d320) at /home/curiousbot/anaconda/gcc/include/c++/bits/shared_ptr_base.h:144
#12 0x00007fffc9aa1c72 in std::_Tuple_impl<0ul, std::shared_ptr<gloo::Algorithm>, std::shared_ptr<char>, std::shared_ptr<char>, std::shared_ptr<std::mutex> >::~_Tuple_impl() ()
   from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#13 0x00007fffc9a9db9f in thd::DataChannelGloo::~DataChannelGloo() () from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#14 0x00007fffc9a9dbc9 in thd::DataChannelGloo::~DataChannelGloo() () from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#15 0x00007fffc9ac5742 in std::unique_ptr<thd::DataChannel, std::default_delete<thd::DataChannel> >::~unique_ptr() () from /home/curiousbot/anaconda/lib/python3.5/site-packages/torch/lib/libTHD.so.1
#16 0x00007ffff6a21ff8 in __run_exit_handlers (status=0, listp=0x7ffff6dac5f8 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true) at exit.c:82
#17 0x00007ffff6a22045 in __GI_exit (status=<optimized out>) at exit.c:104
#18 0x00007ffff6a08837 in __libc_start_main (main=0x400980 <main>, argc=2, argv=0x7fffffffe4f8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe4e8)
    at ../csu/libc-start.c:325
#19 0x00000000004008b9 in _start ()
```

**Comment by jacobgil** [comment]
Hi, I have the same issue.
Is anyone able to use distributed torch with GPUs? Is there a torch version where this works?

**Comment by levyben1** [comment]
Same issue, received similar stack trace when using the pre-build nvidia-docker image of latest pytorch.
Additionally, tried to use the distributed module with the latest pytorch whl file + recompiling from sources and I get a similar stack trace where the exception is raised in gloo::CudaHostPointer<float>::~CudaHostPointer() ().  Head was 66d24c50671fa1886fed65128950240d809b4866 when recompiling from sources.

**Comment by apaszke** [comment]
We know about this problem, but it's really low priority for us, because it has absolutely no effect on run time. It's because global objects are destroyed after CUDA driver libraries are unloaded and this triggers some errors. But, everything up until the program shutdown runs completely fine

**Comment by jcatana** [comment]
Having the same issue. This is a big problem for me running under a batch queuing system. Queue up several jobs to run. The first never completes, hanging indefinitely on this error until manual intervention is performed to kill the process.

**Comment by lliimsft** [comment]
Hi, does it mean all distributed training jobs using Gloo+GPU will fail eventually due to this bug? Is there fix plan for this?

**Comment by vesko-st** [comment]
Same issue here -- this is a problem for me as it terminates the whole program and I want to run other steps after finishing training.

**Comment by apaszke** [comment]
@pankajkumar I disagree. This is only a shutdown-time error, so you can still successfully train your networks in a distributed fashion. It's caused by an incorrect order of destruction of static objects, and I believe this has already been fixed in master.

**Comment by pankajkumar** [comment]
@apaszke Yes but that doesn't mean that there aren't any post processing tasks. Training is getting stopped at the end of every epoch and have to be manually started for each epoch.

**Comment by apaszke** [comment]
If your post processing is defined in Python then it will run. These errors are thrown only right before the whole process exits, after the Python interpreter has already terminated. It shouldn’t affect anything earlier than that so you likely have other problems with your script

**Comment by pankajkumar** [comment]
Okay, thanks for the info. I will check in detail and update.

**Comment by EdwardZhang88** [comment]
@apaszke  I have the same error with distributed training using gloo backend. It happened to both the latest v0.4.0 and also the master 0.5.0a0. Using NCCL2 in v0.4.0 is working fine though. 
terminate called after throwing an instance of 'gloo::EnforceNotMet'
  what():  [enforce fail at /tmp/pip-6uwp_cfg-build/third_party/gloo/gloo/cuda_private.h:40] error == cudaSuccess. 29 vs 0. Error at: /tmp/pip-6uwp_cfg-build/third_party/gloo/gloo/cuda_private.h:40: driver shutting down
@pankajkumar Are you able to get rid of this error?

**Comment by apaszke** [comment]
@EdwardZhang88 while I agree it's not ideal, as I previously mentioned a few times in this thread: this issue is harmless. It only causes an error even *after the Python interpreter exits*, which is only after your script completes successfully.

**Comment by fliman** [comment]
@apaszke you mentioned this bug has been fixed in the master, could you kindly point out which file if it's a simple one we would like to have a quick fix in our local version. The error message is really annoying and fools my logger script all the time.

**Comment by soumith** [comment]
@pietern would know

**Comment by pietern** [comment]
@fliman On master when using THD you can run `torch.distributed.destroy_process_group()` to destruct everything before interpreter shutdown. For 1.0 with c10d all distributed context is captured in Python objects (no more globals) and will be destructed before CUDA shuts down by default (no need for a separate call to destruct it).

**Comment by pietern** [comment]
All objects related to distributed have a Python representation, so they will all be destroyed before the C++ side destructors for global state run. This specific type of error should therefore be gone.
